### PR TITLE
ProbabilisticScorer optimizations

### LINKED ISSUE #1372: Make log optional in `ProbabilisticScorer`
Conversation from slack for context is pasted below, but TL;DR I think we really need to add knobs to `ProbabilisticScorer` to make it more general. While I'm still relatively confident it'll work reasonably well with active probing, its behavior without is pretty bad. At least in large part due to the use of the log on the success probability to linearize it - it simply doesn't penalize nearly enough unless you set the penalty ratio much, much too high, which doesn't help either as then it'll dominate fees. We could consider a different way of combining the scores, but that also sucks as it fails to allow the "larger" score to dominate, which is exactly what you want if you're looking at a path with very high fees or very high failure rates. Ultimately, the only obvious solution is to (optionally) remove the log, which gives us something more akin to the scorer we have today, which has very good results in practice.

While we're at it, we should add a base penalty per-hop option in `ProbabilisticScorer`.

Tagging 106 as ultimately the ProbabilisticScorer is a pretty substantial regression in terms of routing performance on nodes with less data from 105.

```
Matt Corallo
	I'm playing with the new ProbabilisticScorer...some initial observations:
	* It ends up assigning very low penalties - I even doubled the multiplier knob and still see most penalties below 500msat.
	* I didn't want to more-than-double the multiplier knob, given the worst case there starts to look absurd.
	* It seems to be much more willing to take more hops than the old scorer, in line with the lower penalties. This, naturally, seems to cause much higher retry counts.
	* I assume this would change with aggressive probing or high payment volume, but with me just playing around on my own node that isn't the case. (edited) 
	I think we should think about (a) having a more aggressive "default" failure prob and maybe (b) dropping the log() entirely to increase the penalty even if our failure prob shows a low number.
	another idea would be re-introducing a per-hop static cost, irrespective of the failure prob.
Jeffrey Czyz
	what sort of amounts were you sending?
Matt Corallo
	was trying to both "huge" amounts (250k to 2.5m sats) and also one small payment (39k sats)
	not a big N here, mind you, but the old scorer was really aggressive on the limiting hops stuff, and this one just isn't at all, by comparison.
	sadly most of the merchants I regularly pay only take bitpay so I dont have a lot of places I can deposit sats :( (edited) 
	that said, sometimes it works out - just paid satoshis.place 8 sats with a 4 msat fee via 7 hops on the second try (edited) 
	so there is a part of me that wonders if its "doing the right thing" just has a bad default when you dont have enough data.
Jeffrey Czyz
	hmmm... IIUC, it degrades to what c-lightning does when it doesn't have any data. Also note that the default depends on the amount you are sending. Though I suppose that it is the case for the old scorer as of 104 (i.e., because of  overuse_penalty_msat_per_1024th)
Matt Corallo
	sure, but the 104 scorer also had a per-hop static cost, now we may get ~0 costs for a bunch of channels if we have no info, no?
Jeffrey Czyz
	nope, you only get zero penalty if the amount sent is under the lower bound, which starts at zero. So initially there should always be a penalty
Matt Corallo
	sure, but it can be incredibly small, even if not exactly 0
	especially cause the log makes it non-linear, so you can get a very, very, very low penalty for quite some range
Jeffrey Czyz
	Just playing with some numbers to recall how this works. With 1,000 msat multiplier, a payment of 250k sats through a 0.01 BTC channel (25% of the capacity) would give a 125 msat penalty with no information. 50% gives 301 msats, 75% gives 602 msats, 95% gives 1301 msats, and 99% gives 2000 msats, which is also where we artificially limit. So you really need to be using a significant portion of the capacity initially to get a 500 msat penatly. Less if you doubled the multiplier. So for small portions of the channel capacity, it seems the fees will dominate. Just wondering if it would be that bad to increase the multiplier significantly. (edited) 
	-log10((1,000,000,000 + 1 - 990,000,000) / (1,000,000,000 + 1)) * 1,000
Matt Corallo
	the default multiplier is 10k, not 1k msat
Jeffrey Czyz
	ah! right
Matt Corallo
	and I think you mean 250k sat not 250k msat, above.
	also 250k sat is actually kinda high (~$100), common payments are more like 5-20$, afaiu
	so a better analysis would be 25k sats or so
	which is ~100msat in penalty, which will get dwarfed by 1sat base fees (edited) 
Jeffrey Czyz
	yeah, you're right my numbers were off there :sweat_smile:
	or at least my units :slightly_smiling_face:
	anyhow, about 10% of the capacity gives about a 500 msat penalty if I got that right
	with the 10,000 msat multiplier
Matt Corallo
	right, my point is 1m sats is a common channel size, but a common payment amount would be something like 10k sats, which, iiuc, would be a penalty around 50msat
	cause its around 1% of the channel's capacity
	maybe 25k is also common, but that's still ~100msat in penalty
Jeffrey Czyz
	yup, I'm just wondering if increasing the multiplier would suffice, though. If you try to send an amount that is large percentage of the capacity, I would expect a large penalty would be called for
Matt Corallo
	channels are not irregularly unbalanced, I'm pretty dubious the probability is really linear here.
	like sending 95% I'd expect to succeed substantially more than 5% of the time
	so a penalty of like 100 sats seems more than excessive
	which is only a 5x on top of our current penalty multiplier, which seems like it'd be reasonable for the 1-2.5% cases above
	maybe we should add a ton of knobs here and ask cash app to play with them? like option to turn on and off log, mostly.
Jeffrey Czyz
	yeah, that's fair and simple enough change given it wouldn't affect the serialization
Matt Corallo
	right, the really nice thing about the prob. scorer is we can change the scoring pretty substantially without changing what we're storing.

```

### PR DESCRIPTION
**Pull request by jkczyz**
~~Without recent knowledge of channel liquidity balances, `ProbabilisticScorer` tends to give low penalties unless the payment amount is a large portion of a channel's capacity. Provide an option that gives a linear penalty increase as the success probability decreases, which may be useful in such scenarios.~~

~~Additionally, give a base penalty such that shorter paths are preferred over longer ones.~~

Make a few optimizations to `ProbabilisticScorer` including a per hop `base_penalty_msat`. Also, fix a integer overflow that will cause the log approximation to panic.

Closes #1372

### CODE REVIEW

File: lightning/src/routing/scoring.rs [hunk 1]
```diff
@@ -662,11 +704,21 @@ impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiqui
...
+					// Avoid floating-point operations by multiplying the coefficient through:
+					// 2 * liquidity_penalty_multiplier_msat * (1 - success_probability)
+					let coefficient = max_penalty_msat;
+					coefficient.saturating_sub(coefficient.saturating_mul(numerator) / denominator)
+				},
+			}
 		}
 		// Upper bound the penalty to ensure some channel is selected.
-		.min(2 * liquidity_penalty_multiplier_msat)
+		.min(max_penalty_msat)
```
Comment (TheBlueMatt) [inline lightning/src/routing/scoring.rs:721]: Actually, can we drop this? The only case where it should be reachable is the `u64::max_value()` case where we our probability is 0, right?

File: lightning/src/routing/scoring.rs [hunk 2]
```diff
@@ -218,6 +214,22 @@ impl Score for FixedPenaltyScorer {
 	fn payment_path_successful(&mut self, _path: &[&RouteHop]) {}
 }
 
+impl Writeable for FixedPenaltyScorer {
```
Comment (TheBlueMatt) [inline lightning/src/routing/scoring.rs:217]: Should we just drop the writeable entirely?

**Review by TheBlueMatt — COMMENTED** [review]
LGTM. A few comments about docs and some minor nits that don't matter much.

File: lightning/src/routing/scoring.rs [hunk 3]
```diff
@@ -506,19 +518,34 @@ pub struct ProbabilisticScorerUsingTime<G: Deref<Target = NetworkGraph>, T: Time
...
+	/// Default value: [`ProbabilisticScoringCostFunction::NegativeLogSuccessProbability`]
 	///
+	/// [`liquidity_penalty_multiplier_msat`]: Self::liquidity_penalty_multiplier_msat
 	/// [`liquidity_offset_half_life`]: Self::liquidity_offset_half_life
+	pub cost_function: ProbabilisticScoringCostFunction,
+
+	/// A multiplier used in conjunction with [`cost_function`] to determine the channel penalty.
+	///
+	/// The channel penalty is the amount in msats willing to be paid to avoid routing through a
+	/// channel. It is effectively limited to `2 * liquidity_penalty_multiplier_msat`.
```
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:544]: Indeed, let me rephrase now that the base penalty is added.

File: lightning/src/routing/scoring.rs (hunk 1 above)
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:721]: Hmm... I don't think so because it is the difference of two lookups. Would need to increase the lookup table size quite a bit.

Comment (TheBlueMatt) [reply lightning/src/routing/scoring.rs:721]: Ah, right, ugh. Maybe lets just inline this into the `NegativeLogSuccessProbability` case, at least?

File: lightning/src/routing/scoring.rs [hunk 4]
```diff
@@ -537,10 +564,22 @@ pub struct ProbabilisticScoringParameters {
...
+/// A function calculating the cost of routing an amount through a channel.
+///
+/// Costs are calculated in terms of a payment `success_probability`, which is determined by the
+/// effective channel capacity, the payment amount, and knowledge learned from prior successful and
+/// unsuccessful payments. Some cost functions may instead use the `failure_probability`, which is
+/// simply `1.0 - success_probability`. The `success_probability` has a lower bound of `0.01`.
+#[derive(Clone, Copy)]
+pub enum ProbabilisticScoringCostFunction {
+	/// Uses `-log10(success_probability)`, which slowly increases the cost as `success_probability`
+	/// decreases before rapidly increasing.
```
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:576]: Ah, well you did say "less likely to have". 🙂 

Added the suggested comment. Let me know if I got it right or if it could be better phrased.

File: lightning/src/routing/scoring.rs (hunk 1 above)
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:721]: Ideally, this would stay as any cost function should be limited by that. Assuming you have some performance concern? Moved it to the `NegativeLogSuccessProbability` case.

**Review by TheBlueMatt — COMMENTED** [review]
LGTM post-squash.

File: lightning/src/routing/scoring.rs [hunk 5]
```diff
@@ -504,21 +516,36 @@ pub struct ProbabilisticScorerUsingTime<G: Deref<Target = NetworkGraph>, T: Time
...
 	///
+	/// [`liquidity_penalty_multiplier_msat`]: Self::liquidity_penalty_multiplier_msat
 	/// [`liquidity_offset_half_life`]: Self::liquidity_offset_half_life
+	pub cost_function: ProbabilisticScoringCostFunction,
+
+	/// A multiplier used in conjunction with [`cost_function`] to determine the liquidity penalty.
+	///
+	/// The liquidity penalty is effectively limited to `2 * liquidity_penalty_multiplier_msat`.
+	///
+	/// Default value: 10,000 msat
```
Comment (TheBlueMatt) [inline lightning/src/routing/scoring.rs:546]: Note this default is probably much too high for the linear case - I don't think we should change it as long as the `cost_function` default is log, but maybe we want to call out in the docs that it should be changed if the cost function is?

File: lightning/src/routing/scoring.rs [hunk 6]
```diff
@@ -504,21 +516,36 @@ pub struct ProbabilisticScorerUsingTime<G: Deref<Target = NetworkGraph>, T: Time
...
-	/// the success probability is 0.01, effectively limiting the penalty to the range
-	/// `0..=2*liquidity_penalty_multiplier_msat`. The knowledge learned is decayed over time based
-	/// on [`liquidity_offset_half_life`].
+	/// The cost is multiplied by [`liquidity_penalty_multiplier_msat`] to determine the liquidity
+	/// penalty, which is limited to `2 * liquidity_penalty_multiplier_msat`. The cost is based in
+	/// part by the knowledge learned from prior successful and unsuccessful payments. This
+	/// knowledge is decayed over time based on [`liquidity_offset_half_life`].
 	///
-	/// Default value: 10,000 msat
+	/// Default value: [`ProbabilisticScoringCostFunction::NegativeLogSuccessProbability`]
```
Comment (TheBlueMatt) [inline lightning/src/routing/scoring.rs:536]: Do we want to consider changing this? I think log only makes sense if you're doing large payment volume or probing, which I don't think most of our users are?

**Comment by TheBlueMatt** [comment]
Okay, so thinking about the defaults here - lets assume that most channels are 100k-5m in capacity, and that our defaults should target users with 0 learning (or at least all info having decayed) cause they don't have much payment volume. Rene's work concluded that, by far, the failure probability of a hop is, roughly, `amount/capacity` so lets take that as a given, but also that real-world experience seems to indicate that there is some nontrivial probability of a channel being offline that is in excess of that probability. Assume we want to cap the failure penalty for any hop to 20 sats (ie we're willing to pay 20 sats to avoid taking a given path, which is probably about the max we should want for a small payment of, say, 1000 sats).

Lets consider five possible settings -
1) linear with a 20 sats cap,
2) log with a 99% failure cap and a 20 sats cap at 99%,
3) log with a fixed 6.25% failure rate baked in per-hop, with a 10 sats multiplier
4) log with a fixed 12.5% failure rate baked in per-hop, with a 20 sats multiplier as the cap is effectively 87.5,
5) linear with a 20 sats cap with a fixed per-hop penalty of 0.5sats.

Lets consider each setting at 3 hops, 5 hops, and 10 hops for values of 1k, 10k, 50k sats across 100k, and 1m channels.

1k-over-1m (0.1% failure rate)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop |
| --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 0.3%/14.5%/27.3% | 0.06 sats | 0.13 sats | 0.85 sats | 3.51 sats | 1.56 sats |
| 5 hops | 0.5%/23.0%/41.3% | 0.1 sats | 0.22 sats | 1.42 sats | 5.85 sats | 2.6 sats |
| 10 hops | 1.0%/40.8%/65.5% | 0.2 sats | 0.44 sats | 2.85 sats | 11.7 sats | 5.2 sats |

10k-over-1m (1% failure rate, or 10k-over-1m)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop |
| --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 3.0%/16.9%/29.5% | 0.6 sats | 0.13 sats | 0.98 sats | 3.78 sats | 2.1 sats |
| 5 hops | 4.9%/26.6%/44.2% | 1.0 sats | 0.22 sats | 1.63 sats | 6.3 sats | 3.5 sats |
| 10 hops | 9.6%/46.1%/68.8% | 2.0 sats | 0.44 sats | 3.27 sats | 12.6 sats | 7.0 sats |

50k-over-1m (5% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop |
| --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 14.3%/27.1%/38.6% | 3.0 sats | 0.67 sats | 1.55 sats | 5.01 sats | 4.5 sats |
| 5 hops | 22.6%/41.0%/55.6% | 5.0 sats | 1.11 sats | 2.59 sats | 8.35 sats | 7.5 sats |
| 10 hops | 40.1%/65.1%/80.3% | 10.0 sats | 2.23 sats | 5.18 sats | 16.71 sats | 15.0 sats |

10k-over-100k (10% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop |
| --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 27.1%/38.6%/48.8% | 6.0 sats | 1.37 sats | 2.31 sats | 6.64 sats | 7.5 sats |
| 5 hops | 41.0%/55.6%/67.2% | 10.0 sats | 2.29 sats | 3.85 sats | 11.07 sats | 12.5 sats |
| 10 hops | 65.1%/80.3%/89.3% | 20.0 sats | 4.58 sats | 7.7 sats | 22.14 sats | 25.0 sats |

50k-over-100k (50% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop |
| --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 87.5%/90.9%/93.6% | 30.0 sats | 9.03 sats | 10.77 sats | 25.56 sats | 31.5 sats |
| 5 hops | 96.9%/98.2%/99.0% | 50.0 sats | 15.05 sats | 17.95 sats | 42.6 sats | 52.5 sats |
| 10 hops | 99.9%/100.0%/100.0% | 100.0 sats | 30.1 sats | 35.9 sats | 85.19 sats | 105.0 sats |

It think its clear here that log is just completely absurdly low - it reserves half the range for the 90%-99% failure cases, which avoids large penalties through nearly all the data above. Even if we were to double the the penalties applied in log it'd still assign very low penalties, see eg the 10% per-hop failure  table, where even doubling log would get us ~10 sats in penalty for a path that has a 65-80% failure probability. Log 6.25 has a similar issue, although to a smaller extent, with initial per-hop penalties much higher even at low failure rates.

Linear also seems pretty broken, especially in the 10-hops case for lower probability, 0.2 sats for a 10-hop path is unlikely to work out well.

Log 12.5 and linear 0.5-per-hop are both reasonable contenders, but I slightly prefer linear+0.5 here - the high penalties assigned in the 0.1% failure rate table for log 12.5 seem much too conservative even though we end up in almost the same place in the 10% and 50% tables.

**I propose we default to linear with a 0.5sat per-hop penalty**

**Comment by jkczyz** [comment]
I tried log w/ 0.5 sat per hop and a 50 sat multiplier, and it's pretty close to linear w/ 0.5 per-hop, especially at 10% failure probability.

1k-over-1m (0.1% failure rate)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 0.3%/14.5%/27.3% | 0.06 sats | 0.13 sats | 0.85 sats | 3.51 sats | 1.56 sats |2.15 sats | 
| 5 hops | 0.5%/23.0%/41.3% | 0.1 sats | 0.22 sats | 1.42 sats | 5.85 sats | 2.6 sats |3.59 sats | 
| 10 hops | 1.0%/40.8%/65.5% | 0.2 sats | 0.44 sats | 2.85 sats | 11.7 sats | 5.2 sats |7.18 sats | 

10k-over-1m (1% failure rate, or 10k-over-1m)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 3.0%/16.9%/29.5% | 0.6 sats | 0.13 sats | 0.98 sats | 3.78 sats | 2.1 sats |2.15 sats | 
| 5 hops | 4.9%/26.6%/44.2% | 1.0 sats | 0.22 sats | 1.63 sats | 6.3 sats | 3.5 sats |3.59 sats | 
| 10 hops | 9.6%/46.1%/68.8% | 2.0 sats | 0.44 sats | 3.27 sats | 12.6 sats | 7.0 sats |7.18 sats | 

50k-over-1m (5% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 14.3%/27.1%/38.6% | 3.0 sats | 0.67 sats | 1.55 sats | 5.01 sats | 4.5 sats |4.84 sats | 
| 5 hops | 22.6%/41.0%/55.6% | 5.0 sats | 1.11 sats | 2.59 sats | 8.35 sats | 7.5 sats |8.07 sats | 
| 10 hops | 40.1%/65.1%/80.3% | 10.0 sats | 2.23 sats | 5.18 sats | 16.71 sats | 15.0 sats |16.14 sats | 

10k-over-100k (10% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 27.1%/38.6%/48.8% | 6.0 sats | 1.37 sats | 2.31 sats | 6.64 sats | 7.5 sats |8.36 sats | 
| 5 hops | 41.0%/55.6%/67.2% | 10.0 sats | 2.29 sats | 3.85 sats | 11.07 sats | 12.5 sats |13.94 sats | 
| 10 hops | 65.1%/80.3%/89.3% | 20.0 sats | 4.58 sats | 7.7 sats | 22.14 sats | 25.0 sats |27.88 sats | 

50k-over-100k (50% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (10x) | log w/ 12.5% addn'l (20x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 87.5%/90.9%/93.6% | 30.0 sats | 9.03 sats | 10.77 sats | 25.56 sats | 31.5 sats |46.65 sats | 
| 5 hops | 96.9%/98.2%/99.0% | 50.0 sats | 15.05 sats | 17.95 sats | 42.6 sats | 52.5 sats |77.76 sats | 
| 10 hops | 99.9%/100.0%/100.0% | 100.0 sats | 30.1 sats | 35.9 sats | 85.19 sats | 105.0 sats |155.51 sats |

**Comment by TheBlueMatt** [comment]
First of all, I apologize, I had the wrong limit on the logs in the original script, so the 0.1% failure rate table is wrong. Its corrected below (and in the pasted script on the issue).

Hmm, I realize the parameter we're missing in all of this is the payment amount - is 150 sats penalty too much? If you're sending a million sats, probably not, if you're sending 50k sats, probably. I don't think we need to address that there, but in whatever followup we do to pass the scorer more information we should also pass the payment amount and give the `ProbabilisticScorer` a version of the penalty that is multiplied by the payment amount instead of being static.

As for the specific direction to go in here, I added five more options below (swapped the multipliers in the logs with additional probability of failure and added a 30x/40x variant of your proposal).

The 40x variant of your proposal is basically the same as the linear+0.5 proposal for all the numbers that matter (<=10%), and I have to admit I prefer it over 30/50x in the 5% and 10% tables, though of course its all marginal enough to basically be a wash. I hadn't originally thought about doing the linear addition in the log versions because its somewhat nonsensical from a theoretical perspective - why linearize the probability but then add a per-hop penalty - but if it means we can drop the linear option entirely and not branch in the scorer, sure, lets do it.

1k-over-1m (0.1% failure rate)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (20x) | log w/ 12.5% addn'l (10x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (30x) | log w/ 0.5 per hop (40x) | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 0.3%/14.5%/27.3% | 0.06 sats | 0.01 sats | 1.71 sats | 1.75 sats | 1.56 sats |1.54 sats | 1.55 sats | 1.57 sats | 
| 5 hops | 0.5%/23.0%/41.3% | 0.1 sats | 0.02 sats | 2.85 sats | 2.92 sats | 2.6 sats |2.57 sats | 2.59 sats | 2.61 sats | 
| 10 hops | 1.0%/40.8%/65.5% | 0.2 sats | 0.04 sats | 5.7 sats | 5.85 sats | 5.2 sats |5.13 sats | 5.17 sats | 5.22 sats | 

10k-over-1m (1% failure rate, or 10k-over-1m)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (20x) | log w/ 12.5% addn'l (10x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (30x) | log w/ 0.5 per hop (40x) | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 3.0%/16.9%/29.5% | 0.6 sats | 0.13 sats | 1.96 sats | 1.89 sats | 2.1 sats |1.89 sats | 2.02 sats | 2.15 sats | 
| 5 hops | 4.9%/26.6%/44.2% | 1.0 sats | 0.22 sats | 3.27 sats | 3.15 sats | 3.5 sats |3.15 sats | 3.37 sats | 3.59 sats | 
| 10 hops | 9.6%/46.1%/68.8% | 2.0 sats | 0.44 sats | 6.54 sats | 6.3 sats | 7.0 sats |6.31 sats | 6.75 sats | 7.18 sats | 

50k-over-1m (5% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (20x) | log w/ 12.5% addn'l (10x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (30x) | log w/ 0.5 per hop (40x) | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 14.3%/27.1%/38.6% | 3.0 sats | 0.67 sats | 3.11 sats | 2.51 sats | 4.5 sats |3.5 sats | 4.17 sats | 4.84 sats | 
| 5 hops | 22.6%/41.0%/55.6% | 5.0 sats | 1.11 sats | 5.18 sats | 4.18 sats | 7.5 sats |5.84 sats | 6.96 sats | 8.07 sats | 
| 10 hops | 40.1%/65.1%/80.3% | 10.0 sats | 2.23 sats | 10.37 sats | 8.35 sats | 15.0 sats |11.68 sats | 13.91 sats | 16.14 sats | 

10k-over-100k (10% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (20x) | log w/ 12.5% addn'l (10x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (30x) | log w/ 0.5 per hop (40x) | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 27.1%/38.6%/48.8% | 6.0 sats | 1.37 sats | 4.62 sats | 3.32 sats | 7.5 sats |5.62 sats | 6.99 sats | 8.36 sats | 
| 5 hops | 41.0%/55.6%/67.2% | 10.0 sats | 2.29 sats | 7.7 sats | 5.53 sats | 12.5 sats |9.36 sats | 11.65 sats | 13.94 sats | 
| 10 hops | 65.1%/80.3%/89.3% | 20.0 sats | 4.58 sats | 15.4 sats | 11.07 sats | 25.0 sats |18.73 sats | 23.3 sats | 27.88 sats | 

50k-over-100k (50% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (20x) | log w/ 12.5% addn'l (10x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (30x) | log w/ 0.5 per hop (40x) | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 87.5%/90.9%/93.6% | 30.0 sats | 9.03 sats | 21.54 sats | 12.78 sats | 31.5 sats |28.59 sats | 37.62 sats | 46.65 sats | 
| 5 hops | 96.9%/98.2%/99.0% | 50.0 sats | 15.05 sats | 35.9 sats | 21.3 sats | 52.5 sats |47.65 sats | 62.71 sats | 77.76 sats | 
| 10 hops | 99.9%/100.0%/100.0% | 100.0 sats | 30.1 sats | 71.8 sats | 42.6 sats | 105.0 sats |95.31 sats | 125.41 sats | 155.51 sats | 

(75% failure probability)
| --- | expected failure % (Rene/+5%/+10%) | linear | log | log w/ 6.25% addn'l (20x) | log w/ 12.5% addn'l (10x) | linear w/ 0.5 per-hop | log w/ 0.5 per hop (30x) | log w/ 0.5 per hop (40x) | log w/ 0.5 per hop (50x) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 3 hops | 98.4%/99.2%/99.7% | 45.0 sats | 18.06 sats | 43.62 sats | 27.09 sats | 46.5 sats |55.69 sats | 73.75 sats | 91.81 sats | 
| 5 hops | 99.9%/100.0%/100.0% | 75.0 sats | 30.1 sats | 72.7 sats | 45.15 sats | 77.5 sats |92.81 sats | 122.91 sats | 153.01 sats | 
| 10 hops | 100.0%/100.0%/100.0% | 150.0 sats | 60.21 sats | 145.4 sats | 90.31 sats | 155.0 sats |185.62 sats | 245.82 sats | 306.03 sats |

**Comment by jkczyz** [comment]
Ok, so drop the cost function variant and decide on a default multiplier. You're thinking 40x? I think I tried that and 50x on a couple numbers at first and just used 50x in the script because it was a clean 0-100 sat range. But either is fine by me. Could even do a power of two to make it a shift, but then it's either 32 or 64, which is probably either too low or too high.

**Comment by jkczyz** [comment]
> Could even do a power of two to make it a shift, but then it's either 32 or 64, which is probably either too low or too high.

Ah, never mind on this as we are actually using msats.

**Comment by TheBlueMatt** [comment]
> Ok, so drop the cost function variant and decide on a default multiplier. 

Yea, I think that's fine. Of course we basically picked a log probability set which closely hews to linear for most reasonable values, so dunno if it was worth it vs dropping log and using linear to begin with, but, its the code we have, and it is kinda nice that it gets really huge on the upper end of the probability range :man_shrugging: .

> Ah, never mind on this as we are actually using msats.

Well, then we're looking at 64K or 32K, so same thing, basically :p. Yea, sadly I don't think that's granular enough.

**Comment by jkczyz** [comment]
Forced push to kick CI.

> Yea, I think that's fine. Of course we basically picked a log probability set which closely hews to linear for most reasonable values, so dunno if it was worth it vs dropping log and using linear to begin with, but, its the code we have, and it is kinda nice that it gets really huge on the upper end of the probability range 🤷‍♂️ .

I think the benefit of the log is what you mentioned in https://github.com/lightningdevkit/rust-lightning/pull/1375#discussion_r831704754.

**Review by TheBlueMatt — APPROVED** [review]

**Review by valentinewallace — APPROVED** [review]
